README.md
This repo includes a single model (no ensemble) for [toxic classification challange](https://www.kaggle.com/c/jigsaw-multilingual-toxic-comment-classification/overview). The public LB score is ~ 0.93.

The backbone model from Hugginface is finetuned in the notebook _pytorch_model.ipynb_.

The model trained using Google Cloud: https://www.kaggle.com/quanncore/xlm-roberta-large

The notebook for model inference on test data of toxic classification: https://www.kaggle.com/quanncore/pytorch-tpu-inference

Several methods applied in the model:
- Translated training data (from English to other languages) have been used.
- Further fine-tuning has been performed on the validation dataset.
- Data augmentation on the training set has been used.
- Class balancing on a distributed environment has been used.
- TPU is supported.
- A multisample dropout network has been used on the classifier head.
- The hidden states of transformers have been aggregated by attention weights instead of using the last layer.
- Different learning rates for the backbone model and classifier head have been used.
- Stochastic Weight Averaging (SWA) has been used.

Reproduction step of training:
- Create a _credential_ folder and put the Github and Kaggle credential files in JSON format.
- Create a _data_ folder and put the toxic dataset on it. You can change the data paths on the training notebook.
- Create a _model_ folder for storing different training model versions.
- Create a _output_ folder for storing submission files if needed.


changes.txt
-refactor: Refactoring on data_cleaning.py by adding comments and more methods.

scripts/config.py
import torch
from metrics_loss import LabelSmoothing, FocalLosswSmooting
from transformers import get_cosine_with_hard_restarts_schedule_with_warmup


class TrainGlobalConfig:
    num_classes = 2
    num_workers = 0 
    batch_size = 16 
    n_epochs = 4
    # number of epoch validation set is used for finetuning trained model
    n_val_epoch = 3
    lr = 0.1*1e-5
    warmup = 200

    # -------------------
    verbose = True
    verbose_step = 50
    # -------------------

    # --------------------
    step_scheduler = True  # do scheduler.step after optimizer.step
    validation_scheduler = False  # do scheduler.step after validation stage loss
    # SchedulerClass = torch.optim.lr_scheduler.ReduceLROnPlateau
    # scheduler_params = dict(
    #     mode='max',
    #     factor=0.7,
    #     patience=0,
    #     verbose=False, 
    #     threshold=0.0001,
    #     threshold_mode='abs',
    #     cooldown=0, 
    #     min_lr=1e-8,
    #     eps=1e-08
    # )
    # SchedulerClass = torch.optim.lr_scheduler.CyclicLR
    # scheduler_params = dict(
    #     base_lr=lr,
    #     max_lr=1e-4,
    #     step_size_up=2000, 
    #     step_size_down=None, 
    #     mode='triangular2', 
    #     gamma=1.0, 
    #     scale_fn=None, 
    #     scale_mode='cycle', 
    #     cycle_momentum=False, 
    #     base_momentum=0.8, 
    #     max_momentum=0.9, 
    # )
    SchedulerClass = get_cosine_with_hard_restarts_schedule_with_warmup
    scheduler_params = dict(
        num_warmup_steps = warmup,
        num_training_steps = 2000
    )
    # ------------------
    
    # -------------------
    # criterion = FocalLosswSmooting()
    criterion = LabelSmoothing()
    # -------------------
    
    #--------------------
    pseudolabeling_threshold = 0.3
    #--------------------
    
    #--------------------
    # this parameter controls whether we are using shocastic weight avaraging or not
    use_SWA = True
    # this parameter controls whether we are using different lr for backbone model and classifier head
    use_diff_lr = False
    #-------------------

scripts/utility.py
import os
import tqdm
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)


# data related functions
def read_data(dir_paths, languages, english_pipeline=False):
    '''
    Read the related data and return as dataframes.
    If english_pipeline is true, english translated validation test files have been returned.
    '''
    train, valid, test, sub = None, None, None, None
    if 'train_file1' in dir_paths:
        train = pd.read_csv((dir_paths['base_dir']/dir_paths['train_file1']))
    
        if 'train_file2' in dir_paths:
            train2 = pd.read_csv((dir_paths['base_dir']/dir_paths['train_file2']))
            train2 = train2[(train2['toxic'] < 0.2) | (train2['toxic'] > 0.5)]
            train = train.append(train2)
        
        train['lang'] = 'en'
        
        # read translation files if we are not using english pipeline
        if not english_pipeline:
            for lang in languages:
                try:
                    df_path = (dir_paths['base_t_dir']/f'jigsaw-toxic-comment-train-google-{lang}-cleaned.csv')
                    df_lang = pd.read_csv(df_path)
                except:
                    print(f'Translation has not found: {df_path}')
                else:
                    df_lang['lang'] = lang
                    train = train.append(df_lang)
        
        train = train[~train['comment_text'].isna()]
        train = train.drop_duplicates(subset='comment_text')
        train['toxic'] = train['toxic'].round().astype(np.int32)
    
    if english_pipeline:
        valid_translation_columns = ['comment_text_en', 'translated']
        if "val_file_translated1" in dir_paths:
            valid = pd.read_csv((dir_paths['base_t_dir']/dir_paths['val_file_translated1']), index_col='id').drop(['comment_text'], axis=1, errors='ignore')
            valid = valid.rename(columns={col: 'comment_text' for col in valid_translation_columns if col in valid})
        
        if "val_file_translated2" in dir_paths:
            valid2 = pd.read_csv((dir_paths['base_t_dir']/dir_paths['val_file_translated2']), index_col='id').drop(['comment_text'], axis=1, errors='ignore')
            valid2 = valid2.rename(columns={col: 'comment_text' for col in valid_translation_columns if col in valid2})
            valid = valid2 if valid is None else valid.append(valid2)
        
        if valid is not None:
            valid = valid.reset_index(drop=True).drop_duplicates(subset='comment_text')
            valid = valid[~valid['comment_text'].isna()]
            valid['toxic'] = valid['toxic'].round().astype(np.int32)
            valid['lang'] = 'en'
        
        test_translation_columns = ['content_en', 'translated']
        if "test_file_translated1" in dir_paths:
            test1 = pd.read_csv((dir_paths['base_t_dir']/dir_paths['test_file_translated1']), index_col='id').drop(['content'], axis=1, errors='ignore')
            test1 = test1.rename(columns={col: 'content' for col in test_translation_columns if col in test1})
            test1['lang'] = 'en'
            test = [test1]

        if "test_file_translated2" in dir_paths:
            test2 = pd.read_csv((dir_paths['base_t_dir']/dir_paths['test_file_translated2']), index_col='id').drop(['content'], axis=1, errors='ignore')
            test2 = test2.rename(columns={col: 'content' for col in test_translation_columns if col in test2})
            test2['lang'] = 'en'
            test = [test2] if test is None else test + [test2]

    else:
        if 'val_file' in dir_paths:
            valid = pd.read_csv((dir_paths['base_dir']/dir_paths['val_file']), index_col='id')
        if 'test_file' in dir_paths:
            test = pd.read_csv((dir_paths['base_dir']/dir_paths['test_file']), index_col='id')
    
    if 'sub_file' in dir_paths:
        sub = pd.read_csv((dir_paths['base_dir']/dir_paths['sub_file']))
    
    return train, valid, test, sub

def read_external_data(root_path, languages):
    '''
    Read any external data source and return as a single dataframe.
    '''
    DFs, return_df = [], None

    for lang in languages:
        try:
            df_path = (root_path/f'{lang}-external-cleaned.csv')
            if os.path.isfile(df_path):
                df_lang = pd.read_csv(df_path)
                DFs.append(df_lang)
            else:
                print(f'File not found: {df_path}')
        except Exception as e:
            print(f'Error on reading external file: {df_path}')
            raise
    
    if DFs:
        return_df = pd.concat(DFs, ignore_index=True)
    
    return return_df

# encoders
def fast_encode(texts, tokenizer, chunk_size=256, maxlen=512):
    """
    Chunk encode given texts.
    https://www.kaggle.com/xhlulu/jigsaw-tpu-distilbert-with-huggingface-and-keras
    """
    tokenizer.enable_truncation(max_length=maxlen)
    tokenizer.enable_padding(max_length=maxlen)
    all_ids = []
    
    for i in tqdm(range(0, len(texts), chunk_size)):
        text_chunk = texts[i:i+chunk_size].tolist()
        encs = tokenizer.encode_batch(text_chunk)
        all_ids.extend([enc.ids for enc in encs])
    
    return np.array(all_ids)

def regular_encode(texts, tokenizer, maxlen=512, return_mask=False, batch_encode=True):
    '''
    Normal encode given texts.
    '''
    if batch_encode:
        enc_di = tokenizer.batch_encode_plus(
            texts, 
            add_special_tokens=True, 
            pad_to_max_length=True,
            max_length=maxlen
        )
    else:
        enc_di = tokenizer.encode_plus(
            texts, 
            add_special_tokens=True, 
            pad_to_max_length=True,
            max_length=maxlen
        )
    
    if return_mask:
        return np.array(enc_di['input_ids']), np.array(enc_di['attention_mask'])
    
    return np.array(enc_di['input_ids'])

scripts/metrics_loss.py
import numpy as np
import sklearn
import torch

class RocAucMeter(object):
    def __init__(self):
        self.reset()

    def reset(self):
        self.y_true = np.array([0,1])
        self.y_pred = np.array([0.5,0.5])
        self.score = 0

    def update(self, y_true, y_pred):
        y_true = y_true.cpu().numpy().argmax(axis=1)
        y_pred = torch.nn.functional.softmax(y_pred, dim=1).data.cpu().numpy()[:,1]
        self.y_true = np.hstack((self.y_true, y_true))
        self.y_pred = np.hstack((self.y_pred, y_pred))
        self.score = sklearn.metrics.roc_auc_score(self.y_true, self.y_pred, labels=np.array([0, 1]))
    
    @property
    def avg(self):
        return self.score
    
class ThresholdMeter(object):
    '''
    Compute the best threshold value for labelling using validation data. The threshold value could be used in test data pseudolabeling.
    '''
    def __init__(self):
        self.reset()

    def reset(self):
        self.y_true = np.array([0,1])
        self.y_pred = np.array([0.5,0.5])

    def update(self, y_true, y_pred):
        y_true = y_true.cpu().numpy().argmax(axis=1)
        y_pred = torch.nn.functional.softmax(y_pred, dim=1).data.cpu().numpy()[:,1]
        self.y_true = np.hstack((self.y_true, y_true))
        self.y_pred = np.hstack((self.y_pred, y_pred))

    @property
    def find(self):
        # calculate roc curves
        fpr, tpr, thresholds = sklearn.metrics.roc_curve(self.y_true, self.y_pred, pos_label=1)
        # calculate the g-mean for each threshold
        gmeans = np.sqrt(tpr * (1-fpr))
        # locate the index of the largest g-mean
        ix = np.argmax(gmeans)
        
        return thresholds[ix]
                
class AverageMeter(object):
    """Computes and stores the average and current value"""
    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count

        
class LabelSmoothing(torch.nn.Module):
    def __init__(self, smoothing = 0.1):
        super(LabelSmoothing, self).__init__()
        self.confidence = 1.0 - smoothing
        self.smoothing = smoothing

    def forward(self, x, target):
        if self.training:
            x = x.float()
            target = target.float()
            logprobs = torch.nn.functional.log_softmax(x, dim = -1)
            nll_loss = -logprobs * target
            nll_loss = nll_loss.sum(-1)
            smooth_loss = -logprobs.mean(dim=-1)
            loss = self.confidence * nll_loss + self.smoothing * smooth_loss
            return loss.mean()
        else:
            return torch.nn.functional.cross_entropy(x, target)

class FocalLosswSmooting(torch.nn.Module):
    def __init__(self, alpha=0.2, gamma=2., smoothing = 0.1):
        super(FocalLosswSmooting, self).__init__()
        self.alpha = alpha
        self.gamma = gamma
        self.smoothing = smoothing
    
    def forward(self, output, target):
        if self.training:
            output = output.float()
            target = target.float()
            output = torch.sigmoid(output)
            # print(f'target_1 {target}')
            # print(f'output: {output}')
            pos_mask = (target == 1).float()
            neg_mask = (target == 0).float()
            # print(f'pos_mask: {pos_mask}')
            # print(f'neg_mask: {neg_mask}')
            pos_loss = -self.alpha * torch.pow(torch.sub(1.0, output), self.gamma) * torch.log(output) * pos_mask
            neg_loss = -(1 - self.alpha) * torch.pow(output, self.gamma) * \
                       torch.log(torch.sub(1.0, output)) * neg_mask
            # print(f'pos_loss {pos_loss}')
            # print(f'neg_loss {neg_loss}')
    
            neg_loss = neg_loss.sum()
            pos_loss = pos_loss.sum()
            num_pos = pos_mask.view(pos_mask.size(0), -1).sum()
            num_neg = neg_mask.view(neg_mask.size(0), -1).sum()
    
            if num_pos == 0:
                loss = neg_loss
            else:
                loss = pos_loss / num_pos + neg_loss / num_neg
            
            # print(loss)
            # print(-output.sum(dim=-1).mean())
            return (1-self.smoothing)*loss + self.smoothing * -output.sum(dim=-1).mean()
        
        else:
            return torch.nn.functional.cross_entropy(output, target)

scripts/models.py
import warnings
import random
from pathlib import Path

warnings.filterwarnings("ignore")

import time
from datetime import datetime
import numpy as np
import pandas as pd

from transformers import AdamW, get_linear_schedule_with_warmup, get_constant_schedule

import torch
from torch.utils.data import Dataset
from torchcontrib.optim import SWA
from torch.optim import SGD

import torch_xla
import torch_xla.core.xla_model as xm
import torch_xla.distributed.parallel_loader as pl
import torch_xla.distributed.xla_multiprocessing as xmp

from metrics_loss import AverageMeter, RocAucMeter, ThresholdMeter
import albumentation as albm
from utility import regular_encode


def onehot(size, target):
    vec = torch.zeros(size, dtype=torch.float32)
    vec[target] = 1.
    return vec

class DatasetRetriever(Dataset):

    def __init__(self, tokenizer, labels_or_ids, comment_texts, langs, open_subtitles_path=None, maxlen=512, use_train_transforms=False, test=False):
        self.test = test
        self.labels_or_ids = labels_or_ids
        self.comment_texts = comment_texts
        self.langs = langs
        self.tokenizer = tokenizer
        self.maxlen = maxlen
        self.use_train_transforms = use_train_transforms
        if self.use_train_transforms:
            train_transformations = [('exc_mention', 0.95), ('exc_url', 0.95), ('exc_num', 0.95), ('exc_hashtag', 0.95), ('exc_duplicate', 0.95)]
            self.train_trans_chain = albm.get_transforms(train_transformations)
        
            synthesic_transformations = [('synthesic', 0.5)]
            self.synthesic_trans_chain = albm.get_transforms(synthesic_transformations, open_subtitles_path=open_subtitles_path)
            
            shuffle_transformations = [('shuffle', 1)]
            self.shuffle_trans_chain = albm.get_transforms(shuffle_transformations)

    def __len__(self):
        return self.comment_texts.shape[0]

    def __getitem__(self, idx):
        text = self.comment_texts[idx]
        lang = self.langs[idx]
        if self.test is False:
            label = self.labels_or_ids[idx]
            target = onehot(2, label)

        if self.use_train_transforms:
            text, _ = self.train_trans_chain(data=(text, lang))['data']
            tokens, attention_mask = regular_encode(str(text), self.tokenizer, maxlen=self.maxlen, return_mask=True, batch_encode=False)
            token_length = sum(attention_mask)
            if token_length > 0.8*self.maxlen:
                text, _ = self.shuffle_trans_chain(data=(text, lang))['data']
            elif token_length < 60:
                text, _ = self.synthesic_trans_chain(data=(text, label))['data']
            else:
                tokens, attention_mask = torch.tensor(tokens), torch.tensor(attention_mask)
                return target, tokens, attention_mask
        
        tokens, attention_mask = regular_encode(str(text), self.tokenizer, maxlen=self.maxlen, return_mask=True, batch_encode=False)
        tokens, attention_mask = torch.tensor(tokens), torch.tensor(attention_mask)

        if self.test is False:
            return target, tokens, attention_mask
        
        return self.labels_or_ids[idx], tokens, attention_mask

    def get_labels(self):
        return list(np.char.add(self.labels_or_ids.astype(str), self.langs))

class TPUFitter:
    
    def __init__(self, model, device, config, base_model_path='/', model_name='unnamed', model_prefix='roberta', model_version='v1', out_path='/', log_path='/'):
        self.log_path = Path(log_path, 'log').with_suffix('.txt')
        self.log(f'TPUFitter started to initilized.', direct_out=True)
        self.config = config
        self.epoch = 0
        self.base_model_path = base_model_path
        self.model_name = model_name
        self.model_version = model_version
        self.model_path = Path(self.base_model_path, self.model_name, self.model_version)
        
        self.out_path = out_path
        self.node_path = Path(self.out_path, 'node_submissions')
        self.create_dir_structure()

        self.model = model
        self.device = device
        # whether use stochastic weight avaraging
        self.use_SWA = config.use_SWA
        # whether use different lr for backbone and classifier head
        self.use_diff_lr = config.use_diff_lr
        
        self._set_optimizer_scheduler()
        self.criterion = config.criterion
        self.best_score = -1.0
        self.log(f'Fitter prepared. Device is {self.device}', direct_out=True)
    
    def create_dir_structure(self):
        self.node_path.mkdir(parents=True, exist_ok=True)
        self.log(f'**** Directory structure created ****', direct_out=True)
    
    def _set_optimizer_scheduler(self):
        self.log(f'Optimizer and scheduler started to initilized.', direct_out=True)
        def is_backbone(n):
            return 'backbone' in n

        param_optimizer = list(self.model.named_parameters())
        no_decay = ['bias', 'LayerNorm.bias', 'LayerNorm.weight']

        # use different learning rate for backbone transformer and classifier head
        if self.use_diff_lr:
            backbone_lr, head_lr = self.config.lr*xm.xrt_world_size(), self.config.lr*xm.xrt_world_size()*500
            optimizer_grouped_parameters = [
                # {'params': [p for n, p in param_optimizer if not any(nd in n for nd in no_decay)], 'weight_decay': 0.001},
                # {'params': [p for n, p in param_optimizer if any(nd in n for nd in no_decay)], 'weight_decay': 0.0},
                {"params": [p for n, p in param_optimizer if is_backbone(n)], "lr": backbone_lr},
                {"params": [p for n, p in param_optimizer if not is_backbone(n)], "lr": head_lr}
            ]
            self.log(f'Different Learning rate for backbone: {backbone_lr} head:{head_lr}')
        else:
            optimizer_grouped_parameters = [
                {'params': [p for n, p in param_optimizer if not any(nd in n for nd in no_decay)], 'weight_decay': 0.001},
                {'params': [p for n, p in param_optimizer if any(nd in n for nd in no_decay)], 'weight_decay': 0.0},
                ]
        
        try:
            self.optimizer = AdamW(optimizer_grouped_parameters, lr=self.config.lr*xm.xrt_world_size())
            # self.optimizer = SGD(optimizer_grouped_parameters, lr=self.config.lr*xm.xrt_world_size(), momentum=0.9)
        except:
            param_g_1 = [p for n, p in param_optimizer if is_backbone(n)]
            param_g_2 = [p for n, p in param_optimizer if not is_backbone(n)]
            param_intersect = list(set(param_g_1) & set(param_g_2))
            self.log(f'intersect: {param_intersect}', direct_out=True)

        if self.use_SWA:
            self.optimizer = SWA(self.optimizer)
        
        if 'num_training_steps' in self.config.scheduler_params:
            num_training_steps = int(self.config.train_lenght / self.config.batch_size / xm.xrt_world_size() * self.config.n_epochs)
            self.log(f'Number of training steps: {num_training_steps}', direct_out=True)
            self.config.scheduler_params['num_training_steps'] = num_training_steps
        
        self.scheduler = self.config.SchedulerClass(self.optimizer, **self.config.scheduler_params)

    def fit(self, train_loader, validation_loader, n_epochs=None):
        self.log(f'**** Fitting process has been started ****', direct_out=True)
        if n_epochs is None:
            n_epochs = self.config.n_epochs
        
        for e in range(n_epochs):
            if self.config.verbose:
                lr = self.optimizer.param_groups[0]['lr']
                timestamp = datetime.utcnow().isoformat()
                self.log(f'\n{timestamp}\nLR: {lr} \nEpoch:{e}')

            t = time.time()
            para_loader = pl.ParallelLoader(train_loader, [self.device])
            losses, final_scores = self.train_one_epoch(para_loader.per_device_loader(self.device), e)
            
            self.log(f'[RESULT]: Train. Epoch: {self.epoch}, loss: {losses.avg:.5f}, final_score: {final_scores.avg:.5f}, time: {(time.time() - t):.5f}')

            t = time.time()
            para_loader = pl.ParallelLoader(validation_loader, [self.device])
            
            # swap SWA weights for validation
            if self.use_SWA:
                self.log('Swapping SWA weights for validation', direct_out=True)
                self.optimizer.swap_swa_sgd()
            
            losses, final_scores, threshold = self.validation(para_loader.per_device_loader(self.device))
            self.log(f'[RESULT]: Validation. Epoch: {self.epoch}, loss: {losses.avg:.5f}, final_score: {final_scores.avg:.5f}, best_th: {threshold.find:.3f}, time: {(time.time() - t):.5f}')
            # swap back to normal weights to continue training
            if self.use_SWA:
                self.log('Swapping back to original weights for validation', direct_out=True)
                self.optimizer.swap_swa_sgd()
            
            if final_scores.avg > self.best_score:
                self.best_score = final_scores.avg
                self.save('best_model')
                self.log('Best model has been updated', direct_out=True)
                # after one epoch, update SWA model if validation score is increased
                if self.use_SWA:
                    self.optimizer.update_swa()
                    self.log('SWA model weights have been updated', direct_out=True)

            if self.config.validation_scheduler:
                # self.scheduler.step(metrics=final_scores.avg)
                self.scheduler.step()
            
            self.epoch += 1
    
    def run_tuning_and_inference(self, test_loader, validation_loader, validation_tune_loader, n_epochs):
        self.log('******Validation tuning and inference is started*****', direct_out=True)
        self.run_validation_tuning(validation_loader, validation_tune_loader, n_epochs)
        para_loader = pl.ParallelLoader(test_loader, [self.device])
        self.run_inference(para_loader.per_device_loader(self.device))
    
    def run_validation_tuning(self, validation_loader, validation_tune_loader, n_epochs):
        self.log('******Validation tuning is started*****', direct_out=True)
        # self.optimizer.param_groups[0]['lr'] = self.config.lr*xm.xrt_world_size() / (epoch + 1)
        self.fit(validation_tune_loader, validation_loader, n_epochs)
    
    def validation(self, val_loader):
        self.log(f'**** Validation process has been started ****', direct_out=True)
        self.model.eval()
        losses = AverageMeter()
        final_scores = RocAucMeter()
        threshold = ThresholdMeter()

        t = time.time()
        for step, (targets, inputs, attention_masks) in enumerate(val_loader):
            self.log(
                f'Valid Step {step}, loss: ' + \
                f'{losses.avg:.5f}, final_score: {final_scores.avg:.5f}, ' + \
                f'time: {(time.time() - t):.5f}', step=step
            )
            with torch.no_grad():
                inputs = inputs.to(self.device, dtype=torch.long) 
                attention_masks = attention_masks.to(self.device, dtype=torch.long) 
                targets = targets.to(self.device, dtype=torch.float) 

                outputs = self.model(inputs, attention_masks)
                loss = self.criterion(outputs, targets)
                
                batch_size = inputs.size(0)

                final_scores.update(targets, outputs)
                losses.update(loss.detach().item(), batch_size)
                threshold.update(targets, outputs)
        
        return losses, final_scores, threshold

    def train_one_epoch(self, train_loader, epoch):
        self.log(f'**** Epoch training has started: {epoch} ****', direct_out=True)
        self.model.train()

        losses = AverageMeter()
        final_scores = RocAucMeter()
        t = time.time()
        for step, (targets, inputs, attention_masks) in enumerate(train_loader):
            self.log(
                f'Train Step {step}, loss: ' + \
                f'{losses.avg:.5f}, final_score: {final_scores.avg:.5f}, ' + \
                f'time: {(time.time() - t):.5f}', step=step
            )

            inputs = inputs.to(self.device, dtype=torch.long)
            attention_masks = attention_masks.to(self.device, dtype=torch.long)
            targets = targets.to(self.device, dtype=torch.float)

            self.optimizer.zero_grad()

            outputs = self.model(inputs, attention_masks)
            loss = self.criterion(outputs, targets)

            batch_size = inputs.size(0)
            
            final_scores.update(targets, outputs)
            losses.update(loss.detach().item(), batch_size)

            loss.backward()
            xm.optimizer_step(self.optimizer)

            if self.config.step_scheduler:
                self.scheduler.step()
        
        return losses, final_scores

    def run_inference(self, test_loader):
        self.log(f'**** Inference process has been started ****', direct_out=True)
        self.model.eval()
        result = {'id': [], 'toxic': []}
        
        t = time.time()
        for step, (ids, inputs, attention_masks) in enumerate(test_loader):
            self.log(f'Prediction Step {step}, time: {(time.time() - t):.5f}', step=step)

            with torch.no_grad():
                inputs = inputs.to(self.device, dtype=torch.long) 
                attention_masks = attention_masks.to(self.device, dtype=torch.long)
                outputs = self.model(inputs, attention_masks)
                toxics = torch.nn.functional.softmax(outputs, dim=1).data.cpu().numpy()[:,1]

            result['id'].extend(ids.cpu().numpy())
            result['toxic'].extend(toxics)

        result = pd.DataFrame(result)
        print(f'Node path is: {self.node_path}')
        node_count = len(list(self.node_path.glob('*.csv')))
        result.to_csv(self.node_path/f'submission_{node_count}_{datetime.utcnow().microsecond}_{random.random()}.csv', index=False)

    def run_pseudolabeling(self, test_loader, epoch):
        losses = AverageMeter()
        final_scores = RocAucMeter()

        self.model.eval()
        
        t = time.time()
        for step, (ids, inputs, attention_masks) in enumerate(test_loader):

            inputs = inputs.to(self.device, dtype=torch.long) 
            attention_masks = attention_masks.to(self.device, dtype=torch.long)
            outputs = self.model(inputs, attention_masks)
            # print(f'Inputs: {inputs} size: {inputs.size()}')
            # print(f'outputs: {outputs} size: {outputs.size()}')
            toxics = torch.nn.functional.softmax(outputs, dim=1)[:,1]
            toxic_mask = (toxics<=0.4) | (toxics>=0.8)
            # print(attention_masks.size())
            toxics = toxics[toxic_mask]
            inputs = inputs[toxic_mask]
            attention_masks = attention_masks[toxic_mask]
            # print(f'toxics: {toxics.size()}')
            # print(f'inputs: {inputs.size()}')
            if toxics.nelement() != 0:
                targets_int = (toxics>self.config.pseudolabeling_threshold).int()
                targets = torch.stack([onehot(2, target) for target in targets_int])
                # print(targets_int)
                
                self.model.train()
                self.log(
                    f'Pseudolabeling Step {step}, loss: ' + \
                    f'{losses.avg:.5f}, final_score: {final_scores.avg:.5f}, ' + \
                    f'time: {(time.time() - t):.5f}', step=step
                )
    
                targets = targets.to(self.device, dtype=torch.float)
    
                self.optimizer.zero_grad()
    
                outputs = self.model(inputs, attention_masks)
                loss = self.criterion(outputs, targets)
    
                batch_size = inputs.size(0)
                
                final_scores.update(targets, outputs)
                losses.update(loss.detach().item(), batch_size)
    
                loss.backward()
                xm.optimizer_step(self.optimizer)
    
                if self.config.step_scheduler:
                    self.scheduler.step()
    
        self.log(f'[RESULT]: Pseudolabeling. Epoch: {epoch}, loss: {losses.avg:.5f}, final_score: {final_scores.avg:.5f}, time: {(time.time() - t):.5f}')

    def get_submission(self, out_dir):
        submission = pd.concat([pd.read_csv(path) for path in (out_dir/'node_submissions').glob('*.csv')]).groupby('id').mean()
        return submission
    
    def save(self, name):
        self.model_path.mkdir(parents=True, exist_ok=True)
        path = (self.model_path/name).with_suffix('.bin')
        
        if self.use_SWA:
            self.optimizer.swap_swa_sgd()

        xm.save(self.model.state_dict(), path)
        self.log(f'Model has been saved')

    def log(self, message, step=None, direct_out=False):
        if direct_out or self.config.verbose:
            if direct_out or step is None or (step is not None and step % self.config.verbose_step == 0):
                xm.master_print(message)
                with open(self.log_path, 'a+') as logger:
                    xm.master_print(f'{message}', logger)


class ToxicSimpleNNModel(torch.nn.Module):

    def __init__(self, transformer, config):
        super(ToxicSimpleNNModel, self).__init__()
        self.config = config
        self.num_classes = self.config.num_classes
        self.backbone = transformer
        self.dropout = torch.nn.Dropout(p=0.2)
        self.high_dropout = torch.nn.Dropout(p=0.5)
        if self.backbone.config.output_hidden_states:
            print('Multisample dropout is used.')
            n_weights = self.backbone.config.num_hidden_layers + 1
            weights_init = torch.zeros(n_weights).float()
            weights_init.data[:-1] = -3
            self.layer_weights = torch.nn.Parameter(weights_init)
            self.custom_linear_classifier = torch.nn.Linear(
                in_features=self.backbone.config.hidden_size,
                out_features=self.num_classes,
            )
            
            # self.init_weights()
        else:
            self.custom_linear_classifier = torch.nn.Linear(
                in_features=self.backbone.config.hidden_size*2,
                out_features=self.num_classes,
            )

    def forward(self, input_ids, attention_masks):
        bs, seq_length = input_ids.shape
        outputs = self.backbone(input_ids=input_ids, attention_mask=attention_masks)
        
        if self.backbone.config.output_hidden_states:
            hidden_layers = outputs[2]
    
            cls_outputs = torch.stack(
                [self.dropout(layer[:, 0, :]) for layer in hidden_layers], dim=2
            )
            cls_output = (torch.softmax(self.layer_weights, dim=0) * cls_outputs).sum(-1)
    
            # multisample dropout (wut): https://arxiv.org/abs/1905.09788
            logits = torch.mean(
                torch.stack(
                    [self.custom_linear_classifier(self.high_dropout(cls_output)) for _ in range(5)],
                    dim=0,
                ),
                dim=0,
            )
    
            outputs = logits
        
        else:
            seq_x = outputs[0]
            apool = torch.mean(seq_x, 1)
            mpool, _ = torch.max(seq_x, 1)
            x = torch.cat((apool, mpool), 1)
            x = self.dropout(x)
            outputs = self.custom_linear_classifier(x)
        
        return outputs